### PR TITLE
FrontendServerAssetReader

### PR DESCRIPTION
**Pull request by grouma**
- Start building out `FrontendServerAssetReader` for use with the Frontend Server

Towards #865

### CODE REVIEW

**Comment by grouma** [comment]
cc @jonahwilliams

File: dwds/CHANGELOG.md [hunk 1]
```diff
@@ -1,6 +1,7 @@
 ## 1.0.0-dev
 
 - Remove dependency on `package:build_daemon`.
+- Add 'FrontendServerAssetReader` for use with Frontend Server builds.
```
Comment (jakemac53) [inline dwds/CHANGELOG.md]: unmatched quotes

File: dwds/lib/src/readers/frontend_server_asset_reader.dart [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+  final File _mapIncremental;
+  final File _jsonOriginal;
+  final File _jsonIncremental;
+
+  // Map of Dart module server path to source map contents.
+  final _mapContents = <String, String>{};
+
+  bool _haveReadOriginals = false;
+
+  /// Creates a `FrotnendServerAssetReader`.
```
Comment (jakemac53) [inline dwds/lib/src/readers/frontend_server_asset_reader.dart]: `[FrontendServerAssetReader]`

File: dwds/lib/src/readers/frontend_server_asset_reader.dart [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+  /// [outputPath].
+  FrontendServerAssetReader(
+    String outputPath,
+  )   : _mapOriginal = File('$outputPath.map'),
+        _mapIncremental = File('$outputPath.incremental.map'),
+        _jsonOriginal = File('$outputPath.json'),
+        _jsonIncremental = File('$outputPath.incremental.json');
+
+  @override
+  Future<String> dartSourceContents(String serverPath) =>
```
Comment (jakemac53) [inline dwds/lib/src/readers/frontend_server_asset_reader.dart]: This is actually interesting..... this wouldn't be anything specific to the frontend_server.

At the same time we will want an identical abstraction to this one which knows how to serve the JS files.

This class should probably extend an abstract DartSourceAssetReader blass which reads from the FS using a package resolver.

And then this file should also add an additional `jsFileContents(String serverPath)` method which we can plug into our server in webdev.

File: dwds/lib/src/readers/frontend_server_asset_reader.dart [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    return _mapContents[p.withoutExtension(serverPath)];
+  }
+
+  /// Updates the internal caches by reading the Frontend Server output files.
+  ///
+  /// Will only read the incremental files on additional calls.
+  Future<void> updateCaches() async {
+    if (!_haveReadOriginals) {
+      await _updateCaches(_mapOriginal, _jsonOriginal);
+      _haveReadOriginals = true;
```
Comment (jakemac53) [inline dwds/lib/src/readers/frontend_server_asset_reader.dart]: return?

File: dwds/lib/src/readers/frontend_server_asset_reader.dart [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    }
+    await _updateCaches(_mapIncremental, _jsonIncremental);
+  }
+
+  Future<void> _updateCaches(File map, File json) async {
+    if (!(await map.exists() && await json.exists())) return;
```
Comment (jakemac53) [inline dwds/lib/src/readers/frontend_server_asset_reader.dart]: We should consider throwing here (or logging an error at least)

File: dwds/test/readers/frontend_server_asset_reader_test.dart [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+import 'package:test/test.dart';
+
+void main() {
+  FrontendServerAssetReader assetReader;
+  Directory tempFixtures;
+  File jsonOriginal;
+  File mapOriginal;
+
+  Future<void> _createTempFixtures() async {
+    var fixtures = p.absolute(p.relative('test/fixtures', from: p.current));
```
Comment (jakemac53) [inline dwds/test/readers/frontend_server_asset_reader_test.dart]: You shouldn't  need any of the `p.absolute(p.relative(` here. But you should do `p.join('test', 'fixtures')`.

File: dwds/test/readers/frontend_server_asset_reader_test.dart [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+    tempFixtures =
```
Comment (jakemac53) [inline dwds/test/readers/frontend_server_asset_reader_test.dart]: use `Directory.createTemp('dwds_test_fixtures')` instead, or even better use the `test_descriptor` package to manage all of this

File: dwds/test/readers/frontend_server_asset_reader_test.dart [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+  group('FrontendServerAssetReader', () {
+    group('source maps', () {
+      test('can be read', () async {
+        var result =
+            await assetReader.sourceMapContents('web/main.dart.lib.js.map');
+        expect(result, isNotNull);
+      });
+
+      test('are cached', () async {
+        var result =
```
Comment (jakemac53) [inline dwds/test/readers/frontend_server_asset_reader_test.dart]: I would go ahead and just delete these first two lines (previous test covers this, and a read shouldn't affect the caching)

File: dwds/test/readers/frontend_server_asset_reader_test.dart [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+
+        var cachedResult =
+            await assetReader.sourceMapContents('web/main.dart.lib.js.map');
+        expect(cachedResult, isNotNull);
+      });
+
+      test('are null if the path does not exist', () async {
+        var result =
+            await assetReader.sourceMapContents('web/foo.dart.lib.js.map');
+        expect(result, isNull);
```
Comment (jakemac53) [inline dwds/test/readers/frontend_server_asset_reader_test.dart]: With NNBD coming soon we should maybe consider if this is really the API we want or not. The alternative would be to throw, although maybe its easier to deal with the null than the exception?

Comment (grouma) [reply dwds/test/readers/frontend_server_asset_reader_test.dart]: Discussed offline. We will want this field to be nullable in the future. Leaving as is.

File: dwds/lib/src/readers/frontend_server_asset_reader.dart (hunk 4 above)
Comment (grouma) [reply dwds/lib/src/readers/frontend_server_asset_reader.dart]: Changed to an if else

File: dwds/lib/src/readers/frontend_server_asset_reader.dart (hunk 3 above)
Comment (grouma) [reply dwds/lib/src/readers/frontend_server_asset_reader.dart]: Not sure we want to go with the base class approach. I think we'd want to delegate to a `DartSourceAssetReader` or something. I'll cross that hump when I get there though.

File: dwds/test/readers/frontend_server_asset_reader_test.dart (hunk 6 above)
Comment (grouma) [reply dwds/test/readers/frontend_server_asset_reader_test.dart]: I think that is required for Windows testing? It's the pattern we do for the other fixtures, hence why I copied it.

File: dwds/test/readers/frontend_server_asset_reader_test.dart (hunk 7 above)
Comment (grouma) [reply dwds/test/readers/frontend_server_asset_reader_test.dart]: Done.

File: dwds/test/readers/frontend_server_asset_reader_test.dart (hunk 8 above)
Comment (grouma) [reply dwds/test/readers/frontend_server_asset_reader_test.dart]: Done.

File: dwds/test/readers/frontend_server_asset_reader_test.dart (hunk 6 above)
Comment (jakemac53) [reply dwds/test/readers/frontend_server_asset_reader_test.dart]: It shouldn't be necessary, this is just getting passed to a normal File constructor.

**Pull request merged**